src/solve/fulfill.rs

```diff
@@ -346,7 +346,12 @@ impl<'s> Fulfill<'s> {
     /// the outcome of type inference by updating the replacements it provides.
     pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
         let outcome = self.fulfill()?;
-        if outcome.is_complete() && !self.ambiguous {
+
+        if self.cannot_prove {
```

Comment:
Yes and moreover `unify` is called outside of `Fulfill` (see previous comment) so `unify` would still have to change an internal state somehow, so I thought that a flag was lighter.